### ISSUE TITLE
Future compat warning with `.await`ing a `BoxFuture`

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#62312]
The code is the same as in https://github.com/rust-lang/rust/issues/62284:

```rust
#![feature(async_await)]

use std::future::Future;
use std::pin::Pin;

type BoxFuture = Pin<Box<dyn Future<Output = ()>>>;

async fn foo() -> BoxFuture {
    Box::pin(bar()) as _
}

async fn bar() {
    let _ = foo().await;
}
```

This results in:

```
warning[E0381]: use of possibly uninitialized variable: `_`
  --> src/lib.rs:17:13
   |
17 |     let _ = foo().await;
   |             ^^^^^^^^^^^ use of possibly uninitialized value
   |
   = warning: this error has been downgraded to a warning for backwards compatibility with previous releases
   = warning: this represents potential undefined behavior in your code and this warning will become a hard error in the future
   = note: for more information, try `rustc --explain E0729`
```

**Comment by matthewjasper** [comment]
minimized

```rust
fn main() {
    let _ = loop {
        break Box::new(()) as Box<dyn Send>;
    };
}
```

The MIR we generate is broken, so the borrow checker is correct to complain:

```text
        StorageDead(_2); // _2 is uninit after this.
        _1 = move _2 as std::boxed::Box<dyn std::marker::Send> (Pointer(Unsize));
        drop(_2) -> [return: bb7, unwind: bb6]; // What?
```

**Comment by matthewjasper** [comment]
Not an issue on beta.

**Comment by pnkfelix** [comment]
triage: P-high, removing nomination.
